### PR TITLE
Refactor processing

### LINKED ISSUE #1048: Stream processing: Merge StreamProcessor and EventProcessor
No need to distinguish on this level of abstraction. The stream processor can do all those things that the event processor does.

I will rename a couple of things from *event* to *record* along the way.

### LINKED ISSUE #2243: Logstream and Broker are too tightly coupled
Our current stream processing abstraction, in the logstreams module, is hard to test and not optimal.

If we move the `TypedStreamProcessor` in the logstreams module we would be able to write kind of integration tests for the `StreamProcessors` with the `StreamProcessorController`. Currently we are not really able to do this, the current processor tests in the broker tests do not use the real `StreamProcessorController` for example. 
With introducing the error event in https://github.com/zeebe-io/zeebe/issues/2228, we need to test that the `ErrorEvent` are correctly written on unexpected exception. With the current abstraction this is not possible to test.

During this we could think about merging the `EventProcessors` with the `StreamProcessorController` https://github.com/zeebe-io/zeebe/issues/1048.

### LINKED ISSUE #2472: Merge TypedStreamProcessor with the StreamProcessorController
We build a lot of delegates and callbacks because of the splitted modules, e.g.

 * StreamProcessor#onError
    * controller assumes that error event is written and waits until last position is committed, the controller didn't know the record types before    
 * StreamProcessor#getPositionToRecoverFrom
   * on recovery this method is called to get the position from the state - controller was not able to access the state before
  * StreamProcessor#getFailedPosition
    * is called on recovery to check each event whether it is an error record and if so to get the corresponding position to know when to call the onError method again

Probably there a lot more. Since we merged the modules we can now merge the TypedStreamProcessor with the controller and only have one abstraction of the TypedRecordProcessor.  This means the different record processor get the typed records and all the necessary stuff around the processing is done in the controller. This also means we can get rid of some of these callbacks/delegates. 

The controller can also updates the last processed position directly in the state after processing.

related to #1048

### PR DESCRIPTION
**Pull request by ChrisKujawa**
* merge TypedStreamProcessing with StreamProcessorController to
 StreamProcessor
 * StreamProcessor becomes own service, easier to install/create
 * Builders and Contexts are removed or replaced
 * to avoid shared mutable state introduced read only context
 * remove StreamProcessorControl from tests and suspending feature from processor
 * refactor processing and reprocessing tests

closes #2472 
closes #1048 
closes https://github.com/zeebe-io/zeebe/issues/2243

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 55c5e82d1908fe2dcaf1e69866550508f06ff97d
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 77c4f6a62a498ffae5038053201375d714fd4a98
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

File: engine/src/main/java/io/zeebe/engine/processor/TypedRecord.java [hunk 1]
```diff
@@ -22,8 +22,12 @@
 
 public interface TypedRecord<T extends UnpackedObject> {
 
+  long getPosition();
```
Comment (menski) [inline engine/src/main/java/io/zeebe/engine/processor/TypedRecord.java]: I don't think this method should be in the interface. It is not always implement, i.e. `TypedWorkflowInstanceRecord` always returns 0. (https://github.com/zeebe-io/zeebe/pull/2491/files#diff-37ffb731f1b5ec8f32004d4b5ca882ecR51)

Also it is not used in any `TypedRecordProcessor` implementation. It is only used inside the `ProcessingStateMachine` where we have access to the `TypedEventImpl` which contains the position.

File: engine/src/main/java/io/zeebe/engine/processor/TypedRecord.java [hunk 2]
```diff
@@ -22,8 +22,12 @@
...
+
   long getKey();
 
+  long getSourceEventPosition();
```
Comment (menski) [inline engine/src/main/java/io/zeebe/engine/processor/TypedRecord.java]: I don't think this method should be in the interface. It is not always implement, i.e. `TypedWorkflowInstanceRecord` always returns 0. (https://github.com/zeebe-io/zeebe/pull/2491/files#diff-37ffb731f1b5ec8f32004d4b5ca882ecR561)

Also it is only used in test assertions and filters on record streams, where we also have access to the full record.

File: engine/src/main/java/io/zeebe/engine/util/SbeBufferWriterReader.java [hunk 3]
```diff
@@ -85,7 +85,7 @@ public boolean tryWrap(DirectBuffer buffer, int offset, int length) {
   }
 
   public byte[] toBytes() {
-    final byte bytes[] = new byte[getLength()];
+    final byte[] bytes = new byte[getLength()];
```
Comment (menski) [inline engine/src/main/java/io/zeebe/engine/util/SbeBufferWriterReader.java]: :upside_down_face:

**Review by menski — CHANGES_REQUESTED** [review]
LGTM, just a small note on the `TypedRecord` interface

File: engine/src/main/java/io/zeebe/engine/util/SbeBufferWriterReader.java (hunk 3 above)
Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/util/SbeBufferWriterReader.java]: Yeah I tried the intellij refactoring :sweat_smile:

**Comment by Zelldon** [comment]
Thanks for the feedback, totally agree. I applied the review hints.

**Review by menski — APPROVED** [review]

**Comment by menski** [comment]
Thanks :shipit:

**Pull request merged**